### PR TITLE
Draft - Database Redesign - Conversations with 3 or More Users

### PR DESCRIPTION
**Pull request by devNahuelOper**
1. What additional (non-database) changes are needed to implement this feature (we want this to be high level but thorough)?

- Conversation Users: Add a "users" property to the Conversation model. This could be a STRING containing the userId of each user that is part of the conversation i.e. "1, 4, 5". Empty string as a defaultValue. When querying for a particular conversation, we can pass to Conversation.findConversation the IDs of each user in the group, then run a check for conversation.users equaling the ones passed in (if passed in as an array, we can do userArgs.join(",") and then just check for string equality).

- Conversation Add User(s): We could create a Conversation.addUsers method that takes any number of userId arguments, checks for the existence of those users in the database as well as verify a user isn't already part of that conversation (by checking the "users" key of the conversation). userIds that pass the checks get concatenated onto the "users" string and we update the conversation with the new value.

2. How would you update the frontend UI for the new group chat feature?
- In the Chats sidebar, where the names of the other users we are conversing with currently are, I'd have the names of any other conversation participants separated by comments. The way it would show on an iPhone group chat, for instance.
- In the top bar above the ActiveChat component is where I'd place the mechanism for adding more users. I could utilize the already present 3 dots at the right side of the bar, triggering perhaps a dropdown containing the Add User option. Or it could  be a CTA of sorts placed next to the other users' name(s) on that same top bar.
- Currently we display the otherUser avatar beneath the latest message from the currentUser that was read. I'd display the avatars of additional participants right next to where we currently have it, with the number of avatars updating as more users have read the message.

3. How would you update route handlers to reflect the schema change?
- I'd tweak the GET request for fetching all conversations. Currently we use user1Id and user2Id in out 'where' and 'include' arguments in Conversation.findAll. We could have something like: find all the conversations where the "users" string includes our userId from the request param. Once we have made that fetch, we can use the number of users each conversation string has to determine how many User model instances to include for each returned conversation.
When we iterate through those returned convos afterward and determine who the otherUser is based on the convoJSON.user[number] presence, instead of assigning one otherUser object to convoJSON, we can push each existing user[number] object into a convoJSON.otherUsers array.
- I'd add a PUT route for Conversation, for the purpose of updating the "users" property when a user is added to a conversation.

4. How would you update socket event handlers to reflect the schema change?
- In the socket.js file I'd add a listener with an event name such as "add-users-to-conversation". The data it receives would consist of a string of userIds and a conversationId. I'll pass that data to an addUsersToConversation action creator that I'll have added to store/conversations.js.
- In thunkCreators.js I'd have a addUsersToConversation Thunk. It takes a string of userIds and a conversationId, emit a "add-users-to-conversation" event & pass it the same userIds string + conversationId args.

5. How would you update Redux state to reflect the schema change?
- After the new socket event listeners are set up, we can receive the users added to a conversation in our Redux store.
The addUsersToConversation reducer function would be structured somewhat similarly to the function for setting messages read status to the store. We'd receive our string of userIds and conversationId arguments. Find a conversation by the id provided. Create a variable representing a concatenation between the existing users of the convo + the userIds string (I'd check for duplicates here too, avoiding adding an already existing userId). Create a convoCopy with the new users string, then merge this copy with the rest of the conversations state (sans the convo with our given id).

6. In terms of the deployment, you only gave us something to look out for. Can you give us the definitive steps that can achieve the deployment of the new feature?
- Update Conversation model with the "users" property.
- Add PUT route for adding users to a conversation, and adjust the GET route for the newly added "users" property. Perhaps do some API testing on Postman at this stage.
- Integrate socket listeners and Redux store with new Conversation feature.
- On the React side, update all existing components where the additional users would be reflected in the UI i.e. Sidebar Chat, ActiveChat etc.. Perhaps create one or two additional components such as AddUser in the form of a button placed in the ActiveChat top bar & potentially accompanied by a dropdown + search-bar combo for searching all existing users. Another option would be to just update the already present search field to show all users in the database VS only the ones in an active conversation. In this scenario, active convo users would appear at the top of the search with the same UI, while others would appear below & each accompanied by an Add User call to action.
- Test out adding different amounts of users to a conversation under different accounts. Ensure any messages sent within that chat are received by all its users & that the updating of "read" status still operates in the same manner. Ensure all UI updates occur as intended.

### CODE REVIEW

**Review by 1234tgk — COMMENTED** [review]
I would like to ask additional questions about your draft:

1. > `otherUser` property of the conversation could be changed to `otherUsers`, an array of all the other users belonging to a conversation.

    In general, storing an array in a column is a bad idea. It makes querying the individual items inside the array difficult. You can read more about this in [this stackoverflow forum](https://stackoverflow.com/questions/6360739/how-to-store-array-or-multiple-values-in-one-column). Also, we want all conversations to follow a single model for better maintainability. With those in mind, can you give us the new database design?

2. How would you update the **frontend UI** for the new group chat feature?
3. How would you update **route handlers** to reflect the schema change?
4. How would you update **socket event handlers** to reflect the schema change?
5. How would you update **Redux state** to reflect the schema change?
6. In terms of the deployment, you only gave us something to look out for. Can you give us the definitive steps that can achieve the deployment of the new feature?

Please append your answers to your draft, and let us know if it is ready for the second review.

**Review by sonipranjal — APPROVED** [review]
Hi @devNahuelOper, thank you for answering the questions!

PR is approved 🎉 

Please don't merge it since it is only a draft PR.

Some additional questions:

As you have suggested:

> Currently we display the otherUser avatar beneath the latest message from the currentUser that was read. I'd display the avatars of additional participants right next to where we currently have it, with the number of avatars updating as more users have read the message.

Is it a good idea to show all users' avatars below a message? Wouldn't it clutter our UI so much?

> I'd add a PUT route for Conversation, for the purpose of updating the "users" property when a user is added to a conversation.

What will be the relationship between the User and Conversation model? Can you please explain how you will implement this route?

**Pull request merged**
